### PR TITLE
Second part of the PoPCHA example Client

### PR DESCRIPTION
**Pull request by MeKHell**
This PR will add the second part of the example client of the PoPCHA.
It adds the authentication and the proper app part.

The app currently does not verify the signature of the provided tokens.

All the computation is done on the server side, to simplify the app as much as possible.

The `run.sh` script sets up the server's virtual environment and launches it.

`run.sh` optionally takes two arguments (the public server address and the public server port). If **both** arguments are passed, then the script overwrites the config with the provided arguments (this also overwrites the client id).

Here is a screenshot of the app:
image

### CODE REVIEW

**Review by Nazianzenov — APPROVED** [review]
LGTM

File: demo/popcha-client/web/sonar-project.properties
Comment (pierluca) [inline demo/popcha-client/web/sonar-project.properties:11]: you probably don't want to refer to `Go` tests and coverage.
Check the docs for the correct params
https://docs.sonarqube.org/9.9/analyzing-source-code/test-coverage/python-test-coverage/

File: demo/popcha-client/web/sonar-project.properties
Comment (pierluca) [inline demo/popcha-client/web/sonar-project.properties:7]: this looks like it should probably be below :)

File: demo/popcha-client/web/sonar-project.properties
Comment (pierluca) [inline demo/popcha-client/web/sonar-project.properties:4]: You might want to put this along with other python-specific params, to have a clean config file :)

File: demo/popcha-client/web/src/main.py
Comment (pierluca) [inline demo/popcha-client/web/src/main.py:117]: why not move this to the `run()` function ?

**Review by pierluca — COMMENTED** [review]
Just a few minor comments :-)

File: .github/workflows/deploy.yaml
Comment (MeKHell) [reply .github/workflows/deploy.yaml:250]: Ok, I will batch it in the next commit

File: .github/workflows/ci.yaml
Comment (MeKHell) [reply .github/workflows/ci.yaml:309]: I tried to use it first but ran into some trouble, I will give it a look again.

File: demo/popcha-client/web/run.sh
Comment (MeKHell) [reply demo/popcha-client/web/run.sh:10]: If it is empty, it will be generated on launch by the app. Since the configuration would not change, the first time `run.sh` should be called with two arguments  (or the `config.json` should be edited).

File: .github/workflows/ci.yaml
Comment (Tyratox) [reply .github/workflows/ci.yaml:311]: Me neither 😁

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:110]: What if the file doesn't exit?

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:26]: I would try to limit the number of places where you access global variables. It makes functions easier to understand. Here for instance you could accept a dictionary object, modify it and return the modified dictionary object. The caller then updated the global variable / write the data back to disk.

Also I think it would not hurt logging some message saying "Hey I updated your config file"

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:63]: The formatting here is not very readable 😅 Also for some "magic" numbers (44) it would help adding a short comment explaining what the check does.

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:127]:
```suggestion
    provider_strings = [f'{provider.get("lao_id")}@{provider.get("domain")}'
```

Might look a detail but has a different meaning. Is it one string or multiple :)

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:146]: What is the -1 doing here? A default value? Could you also use a named parameter?

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:169]: Looks as if it should be a try / except rather than a "is None" check?

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:195]:
```suggestion
    WARNING: This call is unsafe and allows to easily add new providers. It
```

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:94]:
```suggestion
    Open ID providers, as well as basic config information such as the
```

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:199]:
```suggestion
    is_valid_request = ("domain" in args and "lao_id" in args
```

**Review by Tyratox — COMMENTED** [review]
I added quite a few comments, many of them are regarding the code style. I tried to write a lot of comments because there is no python code in the repo yet and we should decide how we want the code to look like :)

I did not yet get to review the tests though!

File: demo/popcha-client/web/src/main.py
Comment (MeKHell) [reply demo/popcha-client/web/src/main.py:146]: I will, and yes it is a default value.

File: demo/popcha-client/web/data/providers.json
Comment (MeKHell) [reply demo/popcha-client/web/data/providers.json:15]: Yes, this is done on purpose. At the time I did it I did not know the format of the key, I will add correct dummies now.

File: demo/popcha-client/web/src/authentication.py
Comment (MeKHell) [reply demo/popcha-client/web/src/authentication.py:24]: They are not defined by a proper type but only a basic dict. But I will still add it.

File: demo/popcha-client/web/src/authentication.py
Comment (MeKHell) [reply demo/popcha-client/web/src/authentication.py:52]: Done

File: demo/popcha-client/web/src/authentication.py
Comment (MeKHell) [reply demo/popcha-client/web/src/authentication.py:79]: Just checked and both options are valid in PEP8. I checked in this file, and it seems that I always aligned the ending bracket with the argument indentation (see lines 52 and 103)

File: demo/popcha-client/web/src/authentication.py
Comment (MeKHell) [reply demo/popcha-client/web/src/authentication.py:101]: Way better. I did not know `next`

File: demo/popcha-client/web/src/counterApp.py
Comment (MeKHell) [reply demo/popcha-client/web/src/counterApp.py:25]: Update in the function role.

File: demo/popcha-client/web/run.sh
Comment (MeKHell) [reply demo/popcha-client/web/run.sh:10]: After some discussions, since the server does not implement persistency and it is not scheduled to implement it (It is a test server after all), the current behavior should not be a problem. Since the client won't be able to reuse its client id.

File: demo/popcha-client/web/src/main.py
Comment (MeKHell) [reply demo/popcha-client/web/src/main.py]: The run function has been removed. Legacy dev code.

File: demo/popcha-client/web/src/authentication.py
Comment (MeKHell) [reply demo/popcha-client/web/src/authentication.py]: Yes, all of your concerns should be fixed at a later stage. Since it is only an example server and the token creation time is already checked while the token is decoded, I do not think this is that important. In addition, an attacker could easily add new providers to this website through the `/add_provider` get request.

**Comment by MeKHell** [comment]
I will revert the fe1 file to its original state tomorrow morning. If you have any comment you can already review the code.

File: demo/popcha-client/web/src/main.py
Comment (MeKHell) [reply demo/popcha-client/web/src/main.py:141]: Then the process will imedidatly crash with an `FileNotFoundError: [Errno 2] No such file or directory:` error. According to this : [https://stackoverflow.com/questions/15562446/how-to-stop-flask-application-without-using-ctrl-c](https://stackoverflow.com/questions/15562446/how-to-stop-flask-application-without-using-ctrl-c), there is no clean easy way to stop the server on failure, so I think that the current behavior is enough.

File: demo/popcha-client/web/src/authentication.py
Comment (Nazianzenov) [inline demo/popcha-client/web/src/authentication.py:73]: i don't know if it is possible to have a specific error returned if there is a problem (instead of a general "None")

**Review by Nazianzenov — APPROVED** [review]
LGTM, clean code, a small comment on a method signature

File: demo/popcha-client/web/src/authentication.py
Comment (MeKHell) [reply demo/popcha-client/web/src/authentication.py:73]: The idea of this function is to return the ID of the user. I think it is better to return `None` since it is the current user's ID based on the provided arguments.

File: demo/popcha-client/web/src/authentication.py
Comment (Xelowak) [inline demo/popcha-client/web/src/authentication.py:82]: Is there a reason why you omitted the `None` here?

File: demo/popcha-client/web/src/counterApp.py
Comment (Xelowak) [inline demo/popcha-client/web/src/counterApp.py:2]:
```suggestion
This class contains the function related to the application data persitence
```

File: demo/popcha-client/web/src/authentication.py
Comment (Xelowak) [inline demo/popcha-client/web/src/authentication.py:29]: I think it might be nice to add a tiny bit of documentation to know why the format of these parameters is done this way.

File: demo/popcha-client/web/src/counterApp.py
Comment (Xelowak) [inline demo/popcha-client/web/src/counterApp.py:18]: Then why is there a float? Is a user defined by the pair `(str, float)`? It is quite unclear to me 😅 .

**Review by Xelowak — APPROVED** [review]
Well done!
I do not have all the details of your implementation in mind given that I don't know much about PoPCha, but I tried to make useful comments for people like me that don't know how it works in details.

As we discussed today, I'm approving so that you can merge it for tomorrow's release. However. do remember to address those comments and any other comment that may come up  after my review.

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [reply demo/popcha-client/web/src/main.py:141]: I thought it might be nice to show an error but I guess `FileNotFoundError` is explicit enough :)

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:212]: autopep is killing me 😂

File: demo/popcha-client/web/src/main.py
Comment (Tyratox) [inline demo/popcha-client/web/src/main.py:260]:
```suggestion
if __name__ == "__main__":
    on_startup()
```
? :)

**Review by Tyratox — COMMENTED** [review]
LGTM, just a few minor comments :) Great work

**Pull request merged**
